server/ec2spotmanager/CloudProvider/EC2SpotCloudProvider.py

```diff
@@ -0,0 +1,379 @@
+import ssl
+import socket
+import logging
+import datetime
+import botocore
+import boto3
+import boto.ec2
+import boto.exception
+from django.utils import timezone
+from django.conf import settings
+from laniakea.core.providers.ec2 import EC2Manager
+from laniakea.core.userdata import UserData
+from .CloudProvider import CloudProvider, INSTANCE_STATE
+from ..tasks import SPOTMGR_TAG
+from ..common.ec2 import CORES_PER_INSTANCE
+
+
+class EC2SpotCloudProvider(CloudProvider):
+    def __init__(self):
+        self.logger = logging.getLogger("ec2spotmanager")
+
+    def terminate_instances(self, pool_id, instance_ids, terminateByPool=False):
+
+        for region in instance_ids:
+            cluster = EC2Manager(None)
+            try:
+                cluster.connect(region=region, aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+                                aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY)
+            except Exception as msg:
+                raise Exception({"type": "unclassified", "data": msg})
+
+            try:
+                if terminateByPool:
+                    boto_instances = cluster.find(filters={"tag:" + SPOTMGR_TAG + "-PoolId": str(pool_id)})
+
+                    # Data consistency checks
+                    for boto_instance in boto_instances:
+                        # state_code is a 16-bit value where the high byte is
+                        # an opaque internal value and should be ignored.
+                        state_code = boto_instance.state_code & 255
+                        if not ((boto_instance.id in instance_ids[region]) or
+                                (state_code == INSTANCE_STATE['shutting-down'] or
+                                 state_code == INSTANCE_STATE['terminated'])):
+                            self.logger.error("[Pool %d] Instance with EC2 ID %s (status %d) "
+                                              "is not in region list for region %s",
+                                              pool_id, boto_instance.id, state_code, region)
+
+                    cluster.terminate(boto_instances)
+                else:
+                    self.logger.info("[Pool %d] Terminating %s instances in region %s",
+                                     pool_id, len(instance_ids[region]), region)
+                    cluster.terminate(cluster.find(instance_ids=instance_ids[region]))
+            except (boto.exception.EC2ResponseError, boto.exception.BotoServerError, ssl.SSLError, socket.error) as msg:
+                self.logger.exception("[Pool %d] terminate_pool_instances: boto failure: %s", pool_id, msg)
+                return 1
+
+    def start_instances(self, pool_id, config, price_list, blacklisted_zones, image_ids, count=1):
+        images = self._create_laniakea_images(config)
+        # Filter machine sizes that would put us over the number of cores required. If all do, then choose the smallest.
```

Comment:
As mentioned already in `CloudProvider.py`, this kind of business logic should not be in an interface. This logic will apply across providers.

Reply:
this has been removed.